book/scheduling-and-threading.md

```diff
@@ -312,256 +327,241 @@ situation---some devices have more [CPU cores][cores] than others, or are more
+    obj.responseText = body;
+    if (obj.onload)
+        obj.onload(evt);
+}
 ```
 
-Asynchronous rendering
-======================
+And there you have it. With the task machinery and only a few more lines of
+non-plumbing code, we can support lots of new features, and `setTimeout` and
+async `XMLHttpRequest` are only the start.
 
-In order to separate rendering from other tasks and make it into a proper
-pipeline, let's make it asynchronous. Instead of updating rendering right away,
-the browser should set *dirty bits* (a fancy name for a boolean state variable)
-indicating that a particular part of the pipeline needs updating. Then when the
-pipeline is run, it will run the parts indicated by the dirty bits. We'll also
-need some way of
-*scheduling* the rendering pipeline to be updated at a given time in the
-future.
+Rendering pipeline tasks
+========================
 
-First, add one dirty bit to `Tab`:
+Everything in a browser can be considered a task, including rendering. In fact,
+the most important task in a browser is the rendering pipeline---but not just
+for the obvious reason that it's impossible to see web pages that aren't
+rendered. Most of the time spent doing work in a browser is in *rendering
+interactions* with the browser, such as loading, scrolling, clicking and
+typing. All of these interactions require rendering. If you want to make those
+interactions faster and smoother, the very first think you have to do is
+schedule the rendering pipeline, and to achieve that it'll need to be a
+schedulable task.
 
-* `needs_pipeline_update`, indicating that
-the pipeline needs to be re-run.
+On thing that is special about rendering is that it's a "singleton" task. There
+is only one rendering task, and either it's been scheduled or not, but it
+doesn't make sense to schedule it twice at the same time. On the other hand,
+it's totally fine and natural to have many `setTimeout` or `XMLHttpRequest`
+callbacks pending at the same time. The singleton nature of the rendering task
+has to do with there being only one DOM to render.
 
-And two to `TaskRunner`:
+Because it's a singleton, we'll need an additional boolean variable on a `Tab`
+to avoid having two rendering tasks, called `needs_animation_frame`. This
+variable means "a rendering task was already scheduled". Also add a
+`schedule_animation_frame`[^animation-frame] method that adds a new task to
+render, and a new method `run_animation_frame` as the task callback. Note
+how we avoid two scheduled frames with an if statement.
 
-* `display_scheduled`, indicating that a rendering task was already scheduled
-via a `threading.Timer` (this avoids double-running a timer unnecessarily).
-* `needs_animation_frame`, indicating that the event loop should
-run the pipeline ASAP.
+[^animation-frame]: It's called an "animation frame" because sequential
+rendering of different pixels is an animation, and each time you render it's
+one "frame"---like a drawing in a picture frame.
```

Comment:
Ah, the footnote is here. Let's just move this up to where we first use "animation frame"

Reply:
Isn't this the paragraph that first does so?